src/util/threadpool.rs

```diff
@@ -0,0 +1,618 @@
+            }
+            if let Some(task) = task_pool.pop_task() {
+                // to reduce lock's time.
+                task_pool.on_task_started(&task.group_id);
+                return Some(task);
+            }
+            // wait for new task
+            task_pool = cvar.wait(task_pool).unwrap();
+        }
+
+    }
+
+    fn on_task_finished(&self, group_id: &T) {
+        let &(ref lock, _) = &*self.task_pool;
+        let mut task_pool = lock.lock().unwrap();
+        task_pool.on_task_finished(group_id);
+    }
+
+    fn run(&mut self) {
+        // Start the worker.Loop breaks when receive stop message.
+        while let Some(task) = self.get_next_task() {
+            // Since tikv would be down when any panic happens,
+            // we don't need to process panic case here.
+            (task.task)();
+            self.on_task_finished(&task.group_id);
+            self.task_count.fetch_sub(1, AtomicOrdering::SeqCst);
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::{ThreadPool, BigGroupThrottledQueue, Task, ScheduleQueue};
+    use std::thread::sleep;
+    use std::time::Duration;
+    use std::sync::mpsc::channel;
+
+    #[test]
+    fn test_fair_group_for_tasks_with_different_cost() {
+        let name = thd_name!("test_tasks_with_different_cost");
+        let concurrency = 2;
+        let mut task_pool = ThreadPool::new(name, concurrency, BigGroupThrottledQueue::new(1));
+        let (jtx, jrx) = channel();
+        let group_with_big_task = 1001 as u64;
+        let sleep_duration = Duration::from_millis(50);
+        let recv_timeout_duration = Duration::from_secs(2);
+
+        // Push a big task into pool.
+        task_pool.execute(group_with_big_task, move || {
+            sleep(sleep_duration * 10);
+        });
+
+        // Make sure the big task is running.
+        sleep(sleep_duration / 4);
+
+        for group_id in 0..10 {
+            let sender = jtx.clone();
+            task_pool.execute(group_id, move || {
+                sleep(sleep_duration);
+                sender.send(group_id).unwrap();
+            });
+        }
+
+        for _ in 0..10 {
+            let sender = jtx.clone();
+            task_pool.execute(group_with_big_task, move || {
+                sleep(sleep_duration);
+                sender.send(group_with_big_task).unwrap();
+            });
+        }
+
+        // Since a long task of `group_with_big_task` is running,
+        // the other threads shouldn't run any task of `group_with_big_task`.
+        for _ in 0..10 {
+            let group_id = jrx.recv_timeout(recv_timeout_duration).unwrap();
+            assert_ne!(group_id, group_with_big_task);
+        }
+
+        for _ in 0..10 {
+            let second = jrx.recv_timeout(recv_timeout_duration).unwrap();
+            assert_eq!(second, group_with_big_task);
+        }
+        task_pool.stop().unwrap();
+    }
+
+    #[test]
+    fn test_fair_group_for_tasks_with_group_concurrency_on_busy() {
```

Comment:
This test case looks quite similar to the previous `test_fair_group_for_tasks_with_different_cost`. What is the main difference between these two test cases?